### PR TITLE
feat: added authentication in API and search endpoint

### PR DESCRIPTION
**Pull request by k-avy**
feat: added authentication in API and search api
 * Get license with specific search term:
   * field: represents the name of field to be searched
   * search_term: represents the term to be searched
   *  search: represent the algorithm to search with
 * authenticate the API:
      * added the struct for user  
      *  created a user table in the database 
      *  added basic user endpoints:
         * get user using id 
         * get all user
         * create user
      * added basic authentication to all the API endpoints i.e, group of endpoints

### CODE REVIEW

**Comment by avinal** [comment]
Please rebase

**Comment by avinal** [comment]
It seems the rebase was not done correctly, please remove the first commit and rebase again.

File: pkg/authenticate/auth.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+func GetUser(c *gin.Context) {
+	var user models.User
+	id := c.Param("id")
+	err := api.DB.Where("id = ?", id).First(&user).Error
+	if err != nil {
+		er := models.LicenseError{
+			Status:    http.StatusBadRequest,
+			Message:   fmt.Sprintf("invalid request"),
```
Comment (avinal) [inline pkg/authenticate/auth.go:99]: remove sprintf, helpful message

File: pkg/authenticate/auth.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+func AuthenticationMiddleware() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		authHeader := c.GetHeader("Authorization")
+		if authHeader == "" {
+			err := errors.New("authentication failed")
+			er := models.LicenseError{
+				Status:    http.StatusBadRequest,
+				Message:   fmt.Sprintf("Please check your credentials and try again"),
```
Comment (avinal) [inline pkg/authenticate/auth.go:124]: remove sprintf and this is a case of no credentails, indicate that.

File: pkg/authenticate/auth.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	c.JSON(http.StatusOK, res)
+}
+
+func AuthenticationMiddleware() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		authHeader := c.GetHeader("Authorization")
+		if authHeader == "" {
+			err := errors.New("authentication failed")
```
Comment (avinal) [inline pkg/authenticate/auth.go:121]:
```suggestion
```
no need of creating error if you have to get the string again

File: pkg/authenticate/auth.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+			er := models.LicenseError{
+				Status:    http.StatusBadRequest,
+				Message:   fmt.Sprintf("Please check your credentials and try again"),
+				Error:     err.Error(),
```
Comment (avinal) [inline pkg/authenticate/auth.go:125]: write error like "error: no credentials were passed"

File: pkg/authenticate/auth.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+			c.JSON(http.StatusUnauthorized, er)
+			c.Abort()
+			return
+		}
+
+		decodedAuth, err := base64.StdEncoding.DecodeString(strings.TrimPrefix(authHeader, "Basic "))
+		if err != nil {
+			er := models.LicenseError{
+				Status:    http.StatusBadRequest,
+				Message:   fmt.Sprintf("Please check your credentials and try again"),
```
Comment (avinal) [inline pkg/authenticate/auth.go:139]: remove sprintf

File: pkg/authenticate/auth.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+			c.AbortWithStatus(http.StatusBadRequest)
+			return
+		}
+
+		username := auth[0]
+		password := auth[1]
+
+		var user models.User
+		result := api.DB.Where("username = ?", username).First(&user)
+		if result.Error != nil {
```
Comment (avinal) [inline pkg/authenticate/auth.go:161]:
```suggestion
    err := api.DB.Where("username = ?", username).First(&user).Error
	if err != nil {
```

File: pkg/authenticate/auth.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+			er := models.LicenseError{
+				Status:    http.StatusUnauthorized,
+				Message:   fmt.Sprintf("User name not found"),
```
Comment (avinal) [inline pkg/authenticate/auth.go:164]: remove sprintf

File: pkg/authenticate/auth.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+				Error:     result.Error.Error(),
```
Comment (avinal) [inline pkg/authenticate/auth.go:165]:
```suggestion
				Error:     err.Error(),
```

File: pkg/authenticate/auth.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+			c.JSON(http.StatusUnauthorized, er)
+			c.Abort()
+			return
+		}
+
+		// Check if the password matches
+		if user.Userpassword != password {
+			er := models.LicenseError{
+				Status:    http.StatusUnauthorized,
+				Message:   fmt.Sprintf("Incorrect password"),
```
Comment (avinal) [inline pkg/authenticate/auth.go:179]: remove sprintf

File: pkg/authenticate/auth.go [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+				Error:     result.Error.Error(),
```
Comment (avinal) [inline pkg/authenticate/auth.go:180]:
```suggestion
				Error:     err.Error(),
```

**Review by avinal — CHANGES_REQUESTED** [review]
- Only use if you have to create a new string using some variables. Else directly put the string
- when you have a if statement such that the check is only used for the condiiton, combine assignment and if i.e
```go
if err := something.check(); err!= nil {
    // logic
}

// no further use of err
```
- write assignment and if separately if the check is to be reused or there are multiple assignments
```go
value, err := something.Check()
if err != nil {
    // logic
}

// or
val := something.Value()
// check 1
if val.x ... {
    // logic for val.x check
}

// check 2
if val.y ... {
   // logic for val.y check
|
```
keep these guidelines in mind

**Review by avinal — CHANGES_REQUESTED** [review]
You have missed few changes, please fix them

**Review by avinal — APPROVED** [review]
looks good to me, cc @ag4ums

File: cmd/laas/main.go [hunk 11]
```diff
@@ -68,9 +72,14 @@ func main() {
 
 	r := gin.Default()
 	r.NoRoute(api.HandleInvalidUrl)
-	r.GET("/api/licenses", api.GetAllLicense)
-	r.GET("/api/license/:shortname", api.GetLicense)
-	r.POST("/api/license", api.CreateLicense)
-	r.PATCH("/api/license/:shortname", api.UpdateLicense)
+	authorized := r.Group("/")
+	authorized.Use(auth.AuthenticationMiddleware())
+	authorized.GET("/api/license/:shortname", api.GetLicense)
```
Comment (GMishx) [inline cmd/laas/main.go:77]: Why is getting a license requires authorization?

File: cmd/laas/main.go [hunk 12]
```diff
@@ -68,9 +72,14 @@ func main() {
...
+	authorized.POST("/api/license", api.CreateLicense)
+	authorized.PATCH("/api/license/update/:shortname", api.UpdateLicense)
+	authorized.GET("/api/licenses", api.SearchInLicense)
+	r.POST("/api/user", auth.CreateUser)
```
Comment (GMishx) [inline cmd/laas/main.go:81]: And creating a user does not require authentication?

File: cmd/laas/main.go [hunk 13]
```diff
@@ -68,9 +72,14 @@ func main() {
...
-	r.GET("/api/licenses", api.GetAllLicense)
-	r.GET("/api/license/:shortname", api.GetLicense)
-	r.POST("/api/license", api.CreateLicense)
-	r.PATCH("/api/license/:shortname", api.UpdateLicense)
+	authorized := r.Group("/")
+	authorized.Use(auth.AuthenticationMiddleware())
+	authorized.GET("/api/license/:shortname", api.GetLicense)
+	authorized.POST("/api/license", api.CreateLicense)
+	authorized.PATCH("/api/license/update/:shortname", api.UpdateLicense)
+	authorized.GET("/api/licenses", api.SearchInLicense)
```
Comment (GMishx) [inline cmd/laas/main.go:80]: Same here, getting all licenses requires authentication?

File: cmd/laas/main.go [hunk 14]
```diff
@@ -68,9 +72,14 @@ func main() {
...
 	r.NoRoute(api.HandleInvalidUrl)
-	r.GET("/api/licenses", api.GetAllLicense)
-	r.GET("/api/license/:shortname", api.GetLicense)
-	r.POST("/api/license", api.CreateLicense)
-	r.PATCH("/api/license/:shortname", api.UpdateLicense)
+	authorized := r.Group("/")
+	authorized.Use(auth.AuthenticationMiddleware())
+	authorized.GET("/api/license/:shortname", api.GetLicense)
+	authorized.POST("/api/license", api.CreateLicense)
+	authorized.PATCH("/api/license/update/:shortname", api.UpdateLicense)
```
Comment (GMishx) [inline cmd/laas/main.go:79]: Please use consistent plural form for the endpoints. It should be `/api/licenses/update/:shortname` or `/api/licenses/:shortname` not `/api/license/:shortname`.

You can work on it in a separate PR if needed.

File: pkg/api/api.go [hunk 15]
```diff
@@ -188,3 +188,51 @@ func UpdateLicense(c *gin.Context) {
 	c.JSON(http.StatusOK, res)
 
 }
+
+func SearchInLicense(c *gin.Context) {
+	field := c.Query("field")
+	search_term := c.Query("search_term")
```
Comment (GMishx) [inline pkg/api/api.go:194]: I would suggest to read the search term in the body. Keeping it in query limits the size of text someone can search.

Can we discuss it tomorrow @k-avy @avinal @ag4ums ?

**Review by GMishx — CHANGES_REQUESTED** [review]

File: cmd/laas/main.go (hunk 11 above)
Comment (GMishx) [reply cmd/laas/main.go:77]: Ummm... There the endpoint to list all users and specific user is unauthenticated.

Endpoints to Create/modify licenses and every endpoint related to users should be authenticated.

File: cmd/laas/main.go [hunk 16]
```diff
@@ -68,9 +72,15 @@ func main() {
...
 	r.GET("/api/license/:shortname", api.GetLicense)
-	r.POST("/api/license", api.CreateLicense)
-	r.PATCH("/api/license/:shortname", api.UpdateLicense)
+	authorized.POST("/api/license", api.CreateLicense)
+	authorized.PATCH("/api/licenses/update/:shortname", api.UpdateLicense)
+	r.GET("/api/licenses", api.FilterLicense)
+	r.POST("/api/licenses", api.SearchInLicense)
+	r.POST("/api/users", auth.CreateUser)
+	r.GET("/api/users", auth.GetAllUser)
+	r.GET("/api/users/:id", auth.GetUser)
```
Comment (avinal) [inline cmd/laas/main.go:84]: Add auth here

**Review by avinal — CHANGES_REQUESTED** [review]

**Review by avinal — APPROVED** [review]
Looks great to me.

**Review by ag4ums — APPROVED** [review]
LGTM

**Pull request merged**